### PR TITLE
Change rbac roleref to reflect the information we want

### PR DESCRIPTION
**Pull request by deads2k**
@liggitt @ericchiang This is a version of https://github.com/kubernetes/kubernetes/pull/31359 which updates the `RoleRef` to be (I think) the type that we want, with a group, resource, and name.

This is **not** backwards compatible with any existing data.  I'm ok with doing this since rbac was considered alpha, but its something to consider.

If we want this instead, I'll close the previous pull (or update it with this content).

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31543)

### CODE REVIEW

File: pkg/apis/rbac/validation/validation.go [hunk 1]
```diff
@@ -102,11 +102,17 @@ func ValidateRoleBinding(roleBinding *rbac.RoleBinding) field.ErrorList {
...
-			allErrs = append(allErrs, field.Invalid(field.NewPath("roleRef", "namespace"), roleBinding.RoleRef.Namespace, msg))
-		}
+	// TODO allow multiple API groups.  For now, restrict to one, but I can envision other experimental roles in other groups taking
+	// advantage of the binding infrastructure
+	if roleBinding.RoleRef.APIGroup != rbac.GroupName {
+		allErrs = append(allErrs, field.NotSupported(field.NewPath("roleRef", "apiGroup"), roleBinding.RoleRef.APIGroup, []string{rbac.GroupName}))
+	}
+
+	switch roleBinding.RoleRef.Resource {
+	case "role", "clusterrole":
```
Comment (liggitt) [inline pkg/apis/rbac/validation/validation.go]: singular or plural? If we expect someone to be able to use this to build a URL to the resource, I guess we have to do plural, which looks weird :-/

Comment (deads2k) [inline pkg/apis/rbac/validation/validation.go]: > singular or plural? If we expect someone to be able to use this to build a URL to the resource, I guess we have to do plural, which looks weird :-/

I don't feel strongly, plural is fine with me.  I'm more interested in thoughts around compatibility.  I'm inclined to rip off the bandaid....

**Comment by ericchiang** [comment]
As you said it's an alpha so think we're fine to change it, though I'm a little implications of the switch. Does this mean if I have an existing cluster with v1alpha1 types in it I have to purge them before upgrading?

**Comment by deads2k** [comment]
> As you said it's an alpha so think we're fine to change it, though I'm a little implications of the switch. Does this mean if I have an existing cluster with v1alpha1 types in it I have to purge them before upgrading?

You could delete after the fact, but only using a different authorizer.

**Comment by liggitt** [comment]
I like including the group (and defaulting it if empty for `Role` and `ClusterRole`, which makes this a compatible change) and tightening the type to only include the fields we want.

I don't like the switch from `Kind` to `Resource`. That makes this different than other objectrefs, and I have a hard time envisioning a scenario that would justify that pain

**Comment by deads2k** [comment]
> I have a hard time envisioning a scenario that would justify that pain

I can switch it back and continue the charade of confusing what something is with the type you happen to use to serialize it for a merge.

**Comment by deads2k** [comment]
> I don't like the switch from Kind to Resource.

Now it uses Kind

File: pkg/apis/rbac/validation/rulevalidation.go [hunk 2]
```diff
@@ -101,31 +101,24 @@ type ClusterRoleBindingLister interface {
 }
 
 // GetRoleReferenceRules attempts resolve the RoleBinding or ClusterRoleBinding.
-func (r *DefaultRuleResolver) GetRoleReferenceRules(ctx api.Context, roleRef api.ObjectReference, bindingNamespace string) ([]rbac.PolicyRule, error) {
+func (r *DefaultRuleResolver) GetRoleReferenceRules(ctx api.Context, roleRef rbac.RoleRef, bindingNamespace string) ([]rbac.PolicyRule, error) {
 	switch roleRef.Kind {
```
Comment (liggitt) [inline pkg/apis/rbac/validation/rulevalidation.go]: Validate roleRef.APIGroup matches?

File: test/integration/auth/rbac_test.go [hunk 3]
```diff
@@ -303,7 +303,7 @@ func TestRBAC(t *testing.T) {
 						Subjects: []v1alpha1.Subject{
 							{Kind: "User", Name: "pod-reader"},
 						},
-						RoleRef: v1.ObjectReference{Kind: "ClusterRole", Name: "read-pods"},
+						RoleRef: v1alpha1.RoleRef{Resource: "clusterrole", Name: "read-pods"},
```
Comment (liggitt) [inline test/integration/auth/rbac_test.go]: Fix up test data

File: pkg/apis/rbac/types.go [hunk 4]
```diff
@@ -80,6 +80,16 @@ type Subject struct {
...
 }
 
+// RoleRef contains information that points to the role being used
+type RoleRef struct {
+	// APIGroup is the group for the resource being referenced
+	APIGroup string
+	// Kind is the type of resource being referenced
+	Kind string
+	// Name is the name of resource being referenced
+	Name string
```
Comment (liggitt) [inline pkg/apis/rbac/types.go]: implicitly disallowing cross-namespace references by omitting Namespace?

File: pkg/apis/rbac/validation/rulevalidation.go [hunk 5]
```diff
@@ -101,31 +101,24 @@ type ClusterRoleBindingLister interface {
...
 	case "ClusterRole":
 		clusterRole, err := r.clusterRoleGetter.GetClusterRole(api.WithNamespace(ctx, ""), roleRef.Name)
 		if err != nil {
 			return nil, err
 		}
 		return clusterRole.Rules, nil
+
 	default:
-		return nil, fmt.Errorf("unsupported role reference kind: %q", roleRef.Kind)
+		return nil, fmt.Errorf("unsupported role reference resolurce: %q", roleRef.Kind)
```
Comment (liggitt) [inline pkg/apis/rbac/validation/rulevalidation.go]: kind

File: pkg/apis/rbac/types.go (hunk 4 above)
Comment (deads2k) [inline pkg/apis/rbac/types.go]: > implicitly disallowing cross-namespace references by omitting Namespace?

Yes and they were already explicitly disallowed in the evaluation code before.

**Comment by deads2k** [comment]
Are we agreed that we want this over https://github.com/kubernetes/kubernetes/pull/31359 ?

**Comment by ericchiang** [comment]
I'm for this. Having an APIVersion field is already a slightly confusing (though yes, makes things more explicit), and requiring an odd invocation that makes the parser happy doesn't help. If we're going to require the field, I vote for this change.

**Comment by deads2k** [comment]
@liggitt let's break this early.  You good with the change?

**Comment by liggitt** [comment]
LGTM. I don't actually think it's a breaking change, since we're defaulting the group, and cross-namespace was already forbidden

**Comment by CaoShuFeng** [comment]
Hi, can someone help me review this doc:
https://github.com/kubernetes/kubernetes.github.io/pull/1888

**Pull request merged**